### ISSUE TITLE
[feature request] Download multiple packages at the same time

### ISSUE THREAD

**Issue by zutto** [openSUSE/zypper#104]
When doing updates that require downloading of hundreds of packages, it would be great if Zypper supported downloading more than one package at a time.

This would significantly speed up the update process without having to hack a local repository.

**Downsides** of this as of right now is that most of the repository servers still serve pages under http 1.X protocols and might have to throttle clients based on the amount of connections initiated.
But the http servers are moving towards http 2 at very fast rates, and this would mitigate this issue completly if Zypper were to support it as well.

**Upsides** would be that with home connections being at 1Gb/s speeds these days, mobile connections being at ~5-150Mb/s speeds, and 1Gb/s mobile connections (5g) are coming in 3-4 years to the consumer market, the experience would be much better to the user.

Thoughts? Yay's/nay's? Blockers?

**Comment by Qantas94Heavy** [comment]
In terms of the feature itself it should speed things up, though I'm not familiar with how exactly we would implement parallel downloading of packages in libzypp and whether that would require any additional dependencies.

In most cases, download and installation are separate phases, so parallel downloading isn't that different to the current situation, though there may be other issues for other modes (e.g. download and install).

**Comment by avindra** [comment]
I like the way that docker handles the parallel downloading of multiple filesystem layers. It will only pull a (configurable?) maximum number of layers in parallel, and even extract layers in order as they become available.

This would be a huge benefit to the openSUSE ecosystem (esp. for Tumbleweed users who update frequently).

**Comment by liamdawson** [comment]
Connections to my mirrors feel slower in Australia compared to other distros such as Fedora or Ubuntu. I suspect parallel downloads would help make it feel faster, at the least, which is important when I'm doing a 1,200 package `zypper dup` on Tumbleweed. (It generally seems to take a minimum of a second for each package, regardless of size).

**Comment by romulasry** [comment]
Definitely would be nice.

**Comment by romulasry** [comment]
You could use the SHA1 signature on the file to check if it was the same in other repos. That would work.

**Comment by shasheene** [comment]
This is an _incredibly_ important feature. A highly typical OpenSUSE workload on continuous integration servers involve first executing a number of `zypper install pkg_name` steps in a fresh OpenSUSE environment (such as within a container). My connections to an OpenSUSE mirrors via Zypper appeared limited at around 10mbps (regardless of that remote server's utilisation). This wastes a lot of time unnecessarily.

The zypper client should try and utilise all available bandwidth, with the server maintaining the responsibility to manage load-balancing bandwidth between clients.

The official [OpenSUSE mirror list](https://mirrors.opensuse.org) suggests mirrors support HTTP, FTP and rsync. I believe rsync is only used for mirror-to-mirror synchronisation (not by Zypper), but I don't know.

To talk about this feature request in concrete terms, here's a description of the parallelism available at multiple places:
1. Download multiple RPMs concurrently from a single repository
   * This improves total transfer speed if there are a large number of packages
   * Should be easy to implement for HTTP, FTP clients
1. Download a single RPM with multiple connections (starting at different byte offsets) from a single URL (known as "segmented file-transfer")
  * This improves total transfer speed if there is a single large package.
  * Note: now-removed ZYPP_ARIA2C environment variable apparently did this.
1. Download a single RPM with multiple connections (starting at different byte offsets) from multiple repositories (also a segmented file-transfer)

The first two features are unlikely to require much changes. The file list that's presumably generated internally can simply be called by multiple threads. Returning the ability for segmented file-transfer via aria2c is also a great thing to have.

The third feature is likely a huge amount more changes to implement for less net improvement (technically absolutely possible: zypper metadata's repo.xml file means we have the sha1 hashes across N repositories, information so we can generate a URL for each remote repository knowing it's the same file).

**Comment by bzeller** [comment]
Thanks, we already started to think about media backend improvements: 
https://bugzilla.opensuse.org/show_bug.cgi?id=1125470

As for your points: zypper already does use segmented file transfers from multiple different mirrors by utilizing metalinks. However this is only possible with bigger files and only with the official servers. So we already have some of that implemented. However I think the next step really is to go for parallel downloads. This however will require rather big changes in the media backend and is nothing that can happen over night.

The problem here is not running multiple downloads at the same time, but rather the media backend API not being prepared to get multiple file requests at the same time, as well as the frontends (zypper/yast) that are not prepared to get multiple download progress messages at the same time. So while the very core of the feature would be easy to implement, we also need to adapt lots of other code for this to work.

**Comment by harish2704** [comment]
Hi @bzeller , 
if parallel downloading feature requires changes to core components of Zypper, then as a workaround,  it will be helpful if we can simply print the URLs  ( for eg: add an cli option `--print-urls` which will do a dry run without downloading the rpms )

if we can get a list of urls , we can at-least downlad the packages ( in parallel  ) to cache and the re run the Zypper command.

**Comment by Wardrop** [comment]
Agreed. In the age of rolling releases, this is a fairly critical feature. 1400 packages, with a minimum of 2 seconds per package due to latency of each HTTP request, adds up to close to an hour, regardless of the bandwidth you may have at your disposal.

**Comment by bbhoss** [comment]
`zypper install texlive-latex ` to feel real pain on this. It's basically unusable with IPv6 since the mirror it chooses is over 100ms from me. Not sure if that's my ISP's fault or the mirror, but I have to disable IPv6 on the system for major package installations or system upgrades. Parallel downloads would make this issue much less disruptive.

**Comment by awerlang** [comment]
I wrote a POC for downloading packages in parallel. With this script a huge update takes a third of the time (only download).

It doesn't:
* downloads new required/recommended packages
* deals with package installations

https://gist.github.com/awerlang/b792a3f908206a90ad58ba559c5400bb

Improvements welcome.

**Comment by awerlang** [comment]
I improved the POC I posted above, got a 5x speed-up compared to regular `zypper dup --download-only`.

Now it also does:

* downloads new required/recommended packages (zypper dup / inr)
* deals with package installations (zypper in)

**Comment by bzeller** [comment]
Please note that zypper is also doing a checksum check of every file that its downloading. You script is not checking those. Since you are downloading the files directly into the cache you are suggesting to zypper "This file is safe and was checked" so the checksum is not checked again.

**Comment by awerlang** [comment]
@bzeller thanks for the feedback. I'd like to add this missing step, so I've been reading about checksumming the packages. Is it enough to call `rpm -K *.rpm` on each .rpm? Or else, how can I retrieve the checksums? From the repo's primary.xml?

Also, I thought zypper was doing some sort of integrity check on the packages. Because `zypper in` was reporting an empty cache after an aborted download (original size on disk). Resuming the download with the script, then `zypper in` again, the file appears as cached.

**Comment by awerlang** [comment]
I did some testing:

1. Change one byte in a package from tumbleweed-oss: zypper ignores it, aria notices and re-downloads;
2. Replace a package with another from the same repository (packman): zypper ignores it, aria re-downloads (not sure why), but zypper still ignores the package (don't know why);
3. Replace a package with another from another repository: zypper ignores it;

Zypper considered the broken/replaced package as not in cache. `rpm -K` didn't complain about signatures except for test 1.

4. Download a random .rpm: `rpm -K` fails

Anyways, I don't want to hijack the issue. I'm on board on having this feature on zypper as well though.

**Comment by bzeller** [comment]
> @bzeller thanks for the feedback. I'd like to add this missing step, so I've been reading about checksumming the packages. Is it enough to call `rpm --checksig *.rpm` on each .rpm? Or else, how can I retrieve the checksums? From the repo's primary.xml?

Yes, that would be the correct file, but you first would need to parse the repomd.xml to know the exact name of the primary.xml if you pull it from the server, or you'd need to take it from 
/var/cache/zypp/raw/<reponame>/<checksum>-primary.xml.gz. But if you take it from raw cache you'd first need to do a refresh ( thats another thing zypper automatically does that costs some time ).

**Comment by bzeller** [comment]
> I did some testing:
> 
>     1. Change one byte in a package from tumbleweed-oss: zypper ignores it, aria notices and re-downloads;
> 
>     2. Replace a package with another from the same repository (packman): zypper ignores it, aria re-downloads (not sure why), but zypper still ignores the package (don't know why);
> 
>     3. Replace a package with another from another repository: zypper ignores it;
> 
> 
> Zypper considered the broken/replaced package as not in cache. `rpm -K` didn't complain about signatures except for test 1.
> 
>     1. Download a random .rpm: `rpm -K` fails
> 
> 
> Anyways, I don't want to hijack the issue. I'm on board on having this feature on zypper as well though.

Once a file is in the cache, zypper will treat it as "have it already downloaded and was checksummed" iirc. ..  @mlandres knows that in more detail but iirc it would not download it again...

**Comment by mlandres** [comment]
`Package::cachedLocation` considers the checksum, so it should re-load a not matching package. This was introduced because some repos offered the same package (NVRA) but with different content. But basically a file in the cache is expected to be sane. Signature checks performed when downloading are not repeated. If a broken/suspisious file is in the cache we assume the user was notified and accepted using it.

**Comment by awerlang** [comment]
Thanks for the confirmation. Just for principles, can I instruct zypper to ingest these *.rpm from the local dir into zypper cache? That way there's no bypassing zypper safety checks.

**Comment by harish2704** [comment]
Hi,
I just wrote another utility to  print URLs  using zypper. 
https://gist.github.com/harish2704/fdea058e86f1a1e1b806700b061ade2e

Since libzypp and its python bindings are not maintained,  [this script](https://github.com/openSUSE/zypper/issues/104#issuecomment-473514432) can not be used easily .

This utility is a actually two lines of bash script. it derives package url from solver state.

**Comment by awerlang** [comment]
> Hi,
> I just wrote another utility to print URLs using zypper.
> https://gist.github.com/harish2704/fdea058e86f1a1e1b806700b061ade2e
> 
> Since libzypp and its python bindings are not maintained, [this script](https://github.com/openSUSE/zypper/issues/104#issuecomment-473514432) can not be used easily .
> 
> This utility is a actually two lines of bash script. it derives package url from solver state.

The assumption that the architecture is part of the url isn't valid. See Visual Studio Code repository, for example.

**Comment by savek-cc** [comment]
Is anyone currently working on this?

**Comment by bernardosulzbach** [comment]
> Agreed. In the age of rolling releases, this is a fairly critical feature. 1400 packages, with a minimum of 2 seconds per package due to latency of each HTTP request, adds up to close to an hour, regardless of the bandwidth you may have at your disposal.

I would like to emphasize this argument for this feature request. Not having parallel downloads makes network latency a large bottleneck for a task (downloading a few GBs of static data) that should be latency-insensitive.

**Comment by bzeller** [comment]
We are working on easing that problem. It requires lots of restructuring, so it takes time.

**Comment by romulasry** [comment]
Subscribe.

**Comment by D0048** [comment]
Any news on the progress of this?

**Comment by pruebahola** [comment]
any update?

**Comment by mlandres** [comment]
Slowly / steadily. We're currently merging the new  backend code into master. In one of the next libzypp releases it will be possible to enable the backend code via some environment variable. The the new code will take over the downloading.

**But note that this will not enhance anything.**

This is solely for testing and hardening the new downloader. But while it is still driven by the same old serial workflows there will be probably be no noticabel enhancement. This will be the next step todo.

**Comment by Mirppc** [comment]
You are kidding... right?  

Currently mirrorbrain is sending me to the closest mirror with the lowest ping which happens to be running around 20kb/s.  being able to download multiple packages at a single time at 20kb/s will be a HUGE improvement over the current timeoutfest that exist.

More so the option to enable multiple downloads exists in the zypp.conf.  It currently does nothing.  

Now for the new backend.. will this be in the zypper development repository for testing?  If yes then link it here as i will gladly test it.

**Comment by mlandres** [comment]
@Mirppc the download.max_concurrent_connections option in zypp.conf refers to the number of parallel connections we use internally to download the chunks for a single package if we get a metalink file from the server. Those connections however are visible in the zypper.log only, which makes it hard to spot issues if, like in [bsc#1183205](https://bugzilla.opensuse.org/show_bug.cgi?id=1183205#c1), download.opensuse.org delivered poor metalink files. The bugreport contains some hints how to find the connections we use in the log, just in case you want to check it.

**Comment by tecosaur** [comment]
I just wanted to add here, latency is probably the main factor in updates for me. I'm currently doing a tumbleweed `dup` with 7200 packages, at 2s/package that's 240 minutes or 4h.

**Comment by harish2704** [comment]
Currently I m using `dnf` package manager in Tumbleweed ( mainly for distribution upgrades. )
`dup` with zupper is not practical especially if you don't have super-fast internet. 
Previously I had written a script to `pre-download` file ( using python bindings of zypper ) 
But , python bindings of libzypp became unmaintained and then I switched to `dnf`

**Comment by teohhanhui** [comment]
HTTP/2 allows TCP connection reuse. That should be a big win to reduce overhead?

**Comment by tamara-schmitz** [comment]
if this is just about speed. I sped up TW updates noticeably by switching from the official openSUSE downloads to a local mirror: *ftp.fau.de*.
Not only did download speeds for big packages but so did it for smaller ones where connecting may take longer than transferring data. One could investigate there.

**Comment by Mirppc** [comment]
that would be nice if there where mirrors near me.   i let mirrorbrain deal with it even though it goes by lowest ping in my experience.

**Comment by oshanz** [comment]
New zypper HTTP backend:
Another project we have been working on is parallelizing downloads, for 
this a new async downloader backend was implemented.
While it currently won't have massive impact on performance due to the 
frontend code not requesting files asynchronously,
it will do some additional mirror rating and as soon as we update the 
frontend code will bring more benefits.
This can be enabled via setting the env var: ZYPP_MEDIANETWORK=1

https://lists.opensuse.org/archives/list/factory@lists.opensuse.org/thread/YEFSZZY7FRGJPGPWKNCUT3UXQWVENOCL/

**Comment by the-dem** [comment]
Parallel downloading of the metadata would be nice during a refresh. Even if it is one repo at a time (though multiple at once would be even better), it would speed it up slightly. Reading several repos at once to see if they have changed at the beginning would definitely make things quicker.

**Comment by 13werwolf13** [comment]
I will add my opinion.
Parallel updating of metadata from the repositories will really greatly speed up the work, I have more than 40 repositories added to my system and sometimes I wait for metadata updates longer than working with packages.
Parallel downloading of packages will make sense if the user has repositories living on different servers (for example, proprietary nvidia and/or vivaldi repositories are added).

**Comment by mlandres** [comment]
@13werwolf13 right. We meanwhile offer a [tech preview](https://lists.opensuse.org/archives/list/zypp-devel@lists.opensuse.org/thread/YEFSZZY7FRGJPGPWKNCUT3UXQWVENOCL/). By now the preview just contains the new downloader code. As the workflows are still serial the preview is currently mainly for hardening the code. But we are continuously going to enhance the stuff and shipp it to TW and via online updates to all Code15 distros.

In addition to the environment variables mentioned in the announcement, you can permanently enable the preview in `/etc/zypp/zypp.conf` too:
```
[main]
#techpreview.ZYPP_SINGLE_RPMTRANS=1
techpreview.ZYPP_MEDIANETWORK=1
```

**Comment by 13werwolf13** [comment]
> In addition to the environment variables mentioned in the announcement, you can permanently enable the preview in `/etc/zypp/zypp.conf` too:

> techpreview.ZYPP_MEDIANETWORK=1

i try ZYPP_MEDIANETWORK=1 today, 
unfortunately I didn't notice any difference.
my experiment looked like this:
1) ```zypper ref```
2) ```export ZYPP_MEDIA_CURL_IPRESOLVE=4 && zypper dup --allow-vendor-change -d -y```
3) ```export ZYPP_MEDIA_CURL_IPRESOLVE=4 && export ZYPP_MEDIANETWORK=1 && zypper dup --allow-vendor-change -d -y```
unfortunately I did not save the results of the experiment, but the difference was in the margin of error

P.S.: sorry if my poor knowledge of the language will interfere with the understanding of what was said

**Comment by Mirppc** [comment]
> I will add my opinion. Parallel updating of metadata from the repositories will really greatly speed up the work, I have more than 40 repositories added to my system and sometimes I wait for metadata updates longer than working with packages. Parallel downloading of packages will make sense if the user has repositories living on different servers (for example, proprietary nvidia and/or vivaldi repositories are added).

There is also another thing as well.  If the closest repos cap out at 10mbit/s for downloading packages but the person is on fiber or a high speed docsis connection where their network speed to the outside world can be 100mbit/s or more, they can have multiple packages download at the same time maxing out their connection better than doing the downloads one at a time. This also is if the repo that Mirrorbrain or the new alternative send you to a fast repo and not one that has the lowest ping but in fact will cap the download speeds at maybe 250kbit/s (i get this a lot).

**Comment by oshanz** [comment]
> Parallel downloading of packages will make sense if the user has repositories living on different servers (for example, proprietary nvidia and/or vivaldi repositories are added).

@13werwolf13 could you please elaborate on that. do you think parallel downloading from a single server is not going to reduce the total download time? is there any limitations that single server can't deliver parallelly? or any bottleneck at protocol level?

**Comment by 13werwolf13** [comment]
@oshanz
no, of course, in most cases, parallel loading from one server also makes sense, but experience shows that in the case of loading from different servers, this makes much more sense and gives a greater speed increase (although considering how much time has passed since then this may not be relevant an experience).

as an experiment, I would also think about the delivery of packages through something like a torrent or DC ++ but with the amendment to check the package from a central server. repositories and mirrors are good, but if I want to download the `foo-1.2.3` package weighing more than 500 megabytes, then much faster I download it from three neighbors on the network than from a mirror.
I think that those who use for example `cross-avr-gcc` (any version) which weighs a lot and is in the `HOME` repository because of which it does not get on the mirrors will agree with me.
I don't know if I'm alone or not, but I noticed that in the case of opensuse, the more the package weighs, the lower the download speed (and I'm talking about the speed and not about the time spent, which would be logical).

**Comment by Mirppc** [comment]
> > Parallel downloading of packages will make sense if the user has repositories living on different servers (for example, proprietary nvidia and/or vivaldi repositories are added).
> 
> @13werwolf13 could you please elaborate on that. do you think parallel downloading from a single server is not going to reduce the total download time? is there any limitations that single server can't deliver parallely? or any bottleneck at protocol level?

I kinda gave an example of this in my post.  lets say you have 10 100mb packages to download from one repo which has a per download cap of 1mb/s.

With the current method you download one package at a time at 1mb/s.  Depending on speed fluctuations that could take about 17 minutes total.

Now calculate if you have 3 packages downloading at the same time each at 1mb/s per package.  This means the total time spent caching packages would be around 6 minutes.

Now expand this to several thousand packages across oss, non-oss, update oss, update non-oss, packman at minimum and you have quite a bit of time savings

Also there is less likely that foo-1.2.3.rpm got rolled back or updated during the time it took to get to downloading it and that rpm package no longer exists on the repo.  This has happened to me so often.

**Comment by teohhanhui** [comment]
> delivery of packages through something like a torrent … but with the amendment to check the package from a central server

That check would be redundant as they are already sha-256 verified:

https://blog.libtorrent.org/2020/09/bittorrent-v2/

**Comment by 13werwolf13** [comment]
@Mirppc 

I do not know what are the limitations of the mirrors that are chosen by the system for me, but I noticed that if, being behind the same ip, I start updating two machines and more, then the chance of unsuccessful download increases, I suspect that some servers have a limitation on the number of connections with one ip

**Comment by 13werwolf13** [comment]
> > delivery of packages through something like a torrent … but with the amendment to check the package from a central server
> 
> That check would be redundant as they are already sha-256 verified:
> 
> https://blog.libtorrent.org/2020/09/bittorrent-v2/

I clarified just in case, just by the fact that usually when I talk about downloading software from other users, there is always someone who hits himself in the chest and shouts "IT'S NOT SAFE !!!"

**Comment by Mirppc** [comment]
> @Mirppc
> 
> I do not know what are the limitations of the mirrors that are chosen by the system for me, but I noticed that if, being behind the same ip, I start updating two machines and more, then the chance of unsuccessful download increases, I suspect that some servers have a limitation on the number of connections with one ip

This is indeed something that is done and leads to a speed cap to a single IP.  I learned this when dealing with repos and downloading iso's direct from the mirrors hosted by Argonne National Lab and Pacific Northwest National Labs here in the US.

**Comment by 13werwolf13** [comment]
> > @Mirppc
> > I do not know what are the limitations of the mirrors that are chosen by the system for me, but I noticed that if, being behind the same ip, I start updating two machines and more, then the chance of unsuccessful download increases, I suspect that some servers have a limitation on the number of connections with one ip
> 
> This is indeed something that is done and leads to a speed cap to a single IP. I learned this when dealing with repos and downloading iso's direct from the mirrors hosted by Argonne National Lab and Pacific Northwest National Labs here in the US.

Note that in the case of my mirrors this does not affect wget or tools like netboot.xyz, but it does affect zypper and other package managers.

**Comment by dicer** [comment]
As a workaround I use a nightly cronjob to download all the packages but not install them yet. I can then update the system during the day and do not need to download all the packages as they are in the local cache already.
`zypper dup -d --force-resolution --no-confirm --auto-agree-with-licenses`

Maybe this helps people with a really slow download rate or latency.

**Comment by Mirppc** [comment]
> As a workaround I use a nightly cronjob to download all the packages but not install them yet. I can then update the system during the day and do not need to download all the packages as they are in the local cache already. `zypper dup -d --force-resolution --no-confirm --auto-agree-with-licenses`
> 
> Maybe this helps people with a really slow download rate or latency.

for me this would be a quick way to have a broken system due to the multiple repos that have conflicting packages.  this isnt a solution but more of putting the download off until there is a time that the machine isnt being used.  Even then it would be slow AF.  

I have 8 hour streams of me trying to do system migrations as well a updates and the longest part of the process is the download.

Also i have been on 100Mbit/100Mbit or faster internet for a while now so my download speed isnt the issue.

**Comment by dicer** [comment]
> for me this would be a quick way to have a broken system due to the multiple repos that have conflicting packages.

Note the `-d` option which is `--download-only`. The rest of the options just makes sure there are no prompts or failures from conflicts, confirms, or licenses. You still do the resolution when you actually do your `zypper dup` later. The cronjob only fills your local cache.

**Comment by Oisangui** [comment]
Subscribe.

**Comment by bboozzoo** [comment]
> Subscribe.

There's a button on the right hand side, right under 'notifications' which you can click to subscribe.

**Comment by TheKostins** [comment]
Any updates?

**Comment by Yisus7u7** [comment]
any news?

**Comment by chinoto** [comment]
@Yisus7u7 The previously mentioned https://github.com/openSUSE/libzypp/pull/386 was merged. As to how that affects things... 🤷‍♂️

**Comment by mikelpr** [comment]
so, indeed, how does it affect things? I'm on tumbleweed so I suspect my libzypp already has the new media backend, but it doesn't seem like zypper is doing parallel downloads just yet

**Comment by oshanz** [comment]
@mikelpr yes looks like still it's downloading sequentially

I just did an update, there were 291 packages, size of 599 MB
it took 9 minutes to retrieve all the packages. 😑 
According to fast.com my net speed is 440Mbps

**Comment by Mirppc** [comment]
I havent seen any update or change.  If there is i would expect there to be some flag in the zypp or zypper conf like the early experiment in color that was done between the 10.x and 11.x days.  if there is a flag or config change i would like to know so i could test.

currently package download with mirrorsorcer and mirrorcache repos at about 10MiB/s with the main Packman repos downloading between 16-128KiB/s

**Comment by Firstyear** [comment]
Packman is a bad reference for this issue because it's repos are out of our control.

@oshanz It's likely this is related to the issues documented here: https://github.com/Firstyear/opensuse-proxy-cache#technical-details specifically around range requests and how inefficient they are.

**Comment by Mirppc** [comment]
> Packman is a bad reference for this issue because it's repos are out of our control.
> 
> @oshanz It's likely this is related to the issues documented here: https://github.com/Firstyear/opensuse-proxy-cache#technical-details specifically around range requests and how inefficient they are.

not really.. when you can download 3-5 packages from packman at ISDN speeds, things can go faster and you are less likely to need to wait for a packman package to download before something in one of the Opensuse mirrorcache repos gets downloaded.

**Comment by Rush** [comment]
For anybody that feels bad, I have a 5Gbps fiber connection. Zypper makes me feel like we're in the 90's. :-P But the worst is the repo refresh speed? Does anybody know if the new backend work is also addressing that?

**Comment by Firstyear** [comment]
@Rush See: https://github.com/Firstyear/mirrorsorcerer#why-mirrorsorcerer---technical-details to explain the issues.

**Comment by Mirppc** [comment]
> For anybody that feels bad, I have a 5Gbps fiber connection. Zypper makes me feel like we're in the 90's. :-P But the worst is the repo refresh speed? Does anybody know if the new backend work is also addressing that?

There is a solution to this.

see the below 

> @Rush See: https://github.com/Firstyear/mirrorsorcerer#why-mirrorsorcerer---technical-details to explain the issues.

It is miraculous.  on my 500Mbit fiber connection i went from downloading from the Opensuse and OBS repos at 1MiB/s to 10~20MiB/s.

Seriously that thing is Magic!

**Comment by kojq** [comment]
<div align="center">
  <img src="https://user-images.githubusercontent.com/106934605/209165311-b9f1aff8-7aee-4134-a8da-8370541d113f.gif"/>

***Happy 6th Anniversary***
</div>

**Comment by Firstyear** [comment]
This issue is likely too large to take on as a first issue. You may want to look at some other possible contributions first before trying to take this one one.

**Comment by bzeller** [comment]
> Hi!
> 
> I am a new openSUSE user and would like to contribute something to the community. Is this larger "meta" objective still tracked somewhere else, or are there any points where you can help as an open source programmer developer?
> 
> Thanks.

This is a ongoing effort, but it requires rewrites of many parts of the current code base.  Most of the required backend facilities are already in place but the application logic code relies still on the old serial backend, so it needs to be rewritten to a async workflow. So as @Firstyear said its probably a bit too much when getting into the zypp codebase.
You can also look for zypp bugs on https://bugzilla.opensuse.org/buglist.cgi?component=libzypp to find something smaller to work on.

**Comment by Xruptor** [comment]
@bzeller 

I'm a recent convert to Tumbleweed from **Ubuntu->Fedora->Tumbleweed**.  I love the distro but I have major issues here in the **USA->East-> Florida** with the download speeds and overall performance of Zypper.  It's incredibly super duper slow compared to other distributions.  I recently heard about this particular issue ticket from https://www.youtube.com/watch?v=QL5dQmAJPi8&lc He loves the distro as well but is having the same issues I am with updates.

If you don't mind me asking! 😃 What is the current status of this feature?  I'm just curious as the initial ticket was placed back in 2016.  It would be wonderful to hear any update on this particular feature for Zypper.  I think it would be such a boon to the distro especially as it's been in the spotlight as of late.  With the upcoming DNF5 on Fedora being a speedy bullet in terms of perfomance and download speeds, openSUSE Zypper definitely needs some love.

Thanks for such a great distro!

**Comment by Mirppc** [comment]
> @bzeller
> 
> I'm a recent convert to Tumbleweed from **Ubuntu->Fedora->Tumbleweed**. I love the distro but I have major issues here in the **USA->East-> Florida** with the download speeds and overall performance of Zypper. It's incredibly super duper slow compared to other distributions. I recently heard about this particular issue ticket from https://www.youtube.com/watch?v=QL5dQmAJPi8&lc He loves the distro as well but is having the same issues I am with updates.
> 
> If you don't mind me asking! 😃 What is the current status of this feature? I'm just curious as the initial ticket was placed back in 2016. It would be wonderful to hear any update on this particular feature for Zypper. I think it would be such a boon to the distro especially as it's been in the spotlight as of late. With the upcoming DNF5 on Fedora being a speedy bullet in terms of perfomance and download speeds, openSUSE Zypper definitely needs some love.
> 
> Thanks for such a great distro!

So far i havent heard any update on parellel downloads but i do have some reccomendations to help speed things up.  using the download.opensuse.org repos (mirrorbrain powered) i was unable to get beyond 1.2MB/s.  However with Mirrorcache i was able to get 20MB/s.  There is a CDN but that has some problems at times syncing.

So my recommendation would be to check if you are using mirrorcache repo urls .

**Comment by Xruptor** [comment]
> > @bzeller
> > I'm a recent convert to Tumbleweed from **Ubuntu->Fedora->Tumbleweed**. I love the distro but I have major issues here in the **USA->East-> Florida** with the download speeds and overall performance of Zypper. It's incredibly super duper slow compared to other distributions. I recently heard about this particular issue ticket from https://www.youtube.com/watch?v=QL5dQmAJPi8&lc He loves the distro as well but is having the same issues I am with updates.
> > If you don't mind me asking! 😃 What is the current status of this feature? I'm just curious as the initial ticket was placed back in 2016. It would be wonderful to hear any update on this particular feature for Zypper. I think it would be such a boon to the distro especially as it's been in the spotlight as of late. With the upcoming DNF5 on Fedora being a speedy bullet in terms of perfomance and download speeds, openSUSE Zypper definitely needs some love.
> > Thanks for such a great distro!
> 
> So far i havent heard any update on parellel downloads but i do have some reccomendations to help speed things up. using the download.opensuse.org repos (mirrorbrain powered) i was unable to get beyond 1.2MB/s. However with Mirrorcache i was able to get 20MB/s. There is a CDN but that has some problems at times syncing.
> 
> So my recommendation would be to check if you are using mirrorcache repo urls .


Hey there! Thanks for the response 😃  That's a good suggestion and I'm pretty sure I already have it set. The mirrorsorcerer script has me on the "**https://mirrorcache-us-east.opensuse.org**"  repo.  Although it helped a smidgen, it still incredibly slow.  I almost facepalmed when I saw it downloading a 70MB file at a snails pace.

Here's hoping that Zypper gets some love!

**Comment by Firstyear** [comment]
@Xruptor Per https://github.com/Firstyear/mirrorsorcerer/?tab=readme-ov-file#issues-mirrorsorcerer-can-not-prevent please try with ZYPP_MULTICURL=0

**Comment by Rush** [comment]
> It's incredibly super duper slow compared to other distributions.

Indeed. I am using Ubuntu on all my servers and it really pains me to see how vastly better Apt is to Zypper when it comes to speed. Otherwise Tumbleweed is amazing and very very stable. Only a true parallel download solution paired with quick repo servers can truly solve this.

**Comment by bzeller** [comment]
> @Xruptor Per https://github.com/Firstyear/mirrorsorcerer/?tab=readme-ov-file#issues-mirrorsorcerer-can-not-prevent please try with ZYPP_MULTICURL=0

Please do not, multicurl got many improvements over the last few weeks, including zsync and http multirange requests and distribution of chunks over available mirrors up to the max nr of configured connections... e.g. if there is just one mirror all chunks are fetched from there via http multipart.

**Comment by Firstyear** [comment]
They should test this to determine if those changes are actually improving things, or causing issues for them.

**Comment by bzeller** [comment]
> @bzeller
> 
> I'm a recent convert to Tumbleweed from **Ubuntu->Fedora->Tumbleweed**. I love the distro but I have major issues here in the **USA->East-> Florida** with the download speeds and overall performance of Zypper. It's incredibly super duper slow compared to other distributions. I recently heard about this particular issue ticket from https://www.youtube.com/watch?v=QL5dQmAJPi8&lc He loves the distro as well but is having the same issues I am with updates.
> 
> If you don't mind me asking! 😃 What is the current status of this feature? I'm just curious as the initial ticket was placed back in 2016. It would be wonderful to hear any update on this particular feature for Zypper. I think it would be such a boon to the distro especially as it's been in the spotlight as of late. With the upcoming DNF5 on Fedora being a speedy bullet in terms of perfomance and download speeds, openSUSE Zypper definitely needs some love.
> 
> Thanks for such a great distro!

The work is ongoing, we rewriting internal code to support async behavior. This will not only download files in parallel but also do other jobs async. Like checkums, copying files around etc. But since we serve many SUSE and openSUSE releases we need to be careful to not break legacy API, which makes things much more complicated.  Latests work is there openSUSE/libzypp#519

**Comment by pavinjosdev** [comment]
Hey everyone, glad to hear async work is quite a bit underway. The PR seems like it's fully baked 🎂 and the only thing holding it back is breaking changes to the API? 
Hope it can be added to Tumbleweed, if not as zypper then perhaps as a new package zypper-async or something. Sort of like dnf and dnf5.
I wrote a poor man's version using python async and chroot concepts for refreshing and downloading packages in parallel:
https://github.com/pavinjosdev/zypperoni

The installation is still quite slow, this is the only thing that I can't do safely now. Hoping to see async zypper in Tumbleweed soon 🙂🤞

**Comment by bzeller** [comment]
> Hey everyone, glad to hear async work is quite a bit underway. The PR seems like it's fully baked 🎂 and the only thing holding it back is breaking changes to the API? 

That was holding it back for a long time because I was reluctant to duplicate lots of code just to have sync and async versions of it. The new approach with what we call pipelines now allows us to compile the same code as sync and async variants and only the top level API needs to be implemented for async.  

And also we first had to implement the foundation for async code.

>Hope it can be added to Tumbleweed, if not as zypper then perhaps as a new package zypper-async or something. Sort of like dnf and dnf5.

The point is that we can add it to as many releases as possible without huge maintenance overhead, we just have 2 people working on zypp. Ofc Tumbleweed will get it. We have not yet fully decided if we will add a new frontend... something like zypp vs zypper maybe. That depends how bad we'd break behavior once zypper would start to do async workflows. 

Once I have something working we currently consider to release a async-refresh tool that ties into zypper and can be installed if someone wants to testrun the new functionality, e.g. by executing `zypper refresh-async`

**Comment by pavinjosdev** [comment]
> Once I have something working we currently consider to release a async-refresh tool that ties into zypper and can be installed if someone wants to testrun the new functionality, e.g. by executing `zypper refresh-async`

Would this plugin to zypper affect the default repo refresh while doing `zypper <info | search>`? Or perhaps that would only happen with the major async zypper addition itself?

Thanks for all your work. It has a huge impact and certainly positive on so many users 😀

**Comment by bzeller** [comment]
> Would this plugin to zypper affect the default repo refresh while doing `zypper <info | search>`? Or perhaps that would only happen with the major async zypper addition itself?

No initially it would not have impact on the normal zypper commands. That would be just a way for people to test bleeding edge async code.  If we figure out a way to make zypper async under the hood without breaking behavior too much then we'll do that... Otherwise we will opt for a new frontend , just like dnf did.

**Comment by awerlang** [comment]
Those benchmarking zypper can use this PoC announced at https://github.com/openSUSE/zypper/issues/104#issuecomment-607520737
(not a benchmarking tool, but a different parallel downloads implementation which works)

I have been using my PoC for a few hundreds dups successfully. Compared with zypper dup today:

```
baseurl=https://cdn.opensuse.org/tumbleweed/repo/oss
        https://mirrorcache-br.opensuse.org/tumbleweed/repo/oss

$ time zypper --non-interactive dup --download-only 
real    14m43.982s
user    0m0.193s
sys     0m0.388s
 
$ time zypper-download dup
real    1m14.316s
user    0m20.734s
sys     0m9.177s
```

I retried with a single repo in baseurl to similar results.

**Comment by pavinjosdev** [comment]
> I love the name for it! That was quite clever and made me chuckle! 😄 I'm going to give this script and try and see if there are any improvements. I think downloading and then applying the updates afterwards wouldn't be so bad. Especially if it's not going to take close to an hour to download.

Thanks! Downloads are pretty fast with zypperoni, `zypper dup` doesn't have to download anything as all the packages it needs would already be in the cache. Installation of around 2000 packages for a routine upgrade still takes around 30 minutes on a modest VM though (4 CPU, 4G RAM) 🥲. This can't be handled safely by external programs I think, at least with zypperoni it can only make writes to `/var/cache/zypp`. Everything else is read-only to avoid any chance of clobbering up the system.

**Comment by bzeller** [comment]
> > I love the name for it! That was quite clever and made me chuckle! 😄 I'm going to give this script and try and see if there are any improvements. I think downloading and then applying the updates afterwards wouldn't be so bad. Especially if it's not going to take close to an hour to download.
> 
> Thanks! Downloads are pretty fast with zypperoni, `zypper dup` doesn't have to download anything as all the packages it needs would already be in the cache. Installation of around 2000 packages for a routine upgrade still takes around 30 minutes on a modest VM though (4 CPU, 4G RAM) 🥲. This can't be handled safely by external programs I think, at least with zypperoni it can only make writes to `/var/cache/zypp`. Everything else is read-only to avoid any chance of clobbering up the system.

We have a improvement for installing packages already pending, it is currently disabled for some minor incompatibility reasons but many already use it every day without problems. 

Either put `techpreview.ZYPP_SINGLE_RPMTRANS=1` into your zypp.conf or try it by setting it as environment variable when executing zypper: `env ZYPP_SINGLE_RPMTRANS=1 zypper dup` . This speeds up big updates by running it as a single rpm transaction.

However it is NOT possible to install/upgrade packages in parallel... rpm does not support that.

**Comment by pavinjosdev** [comment]
> Either put `techpreview.ZYPP_SINGLE_RPMTRANS=1` into your zypp.conf or try it by setting it as environment variable when executing zypper: `env ZYPP_SINGLE_RPMTRANS=1 zypper dup` . This speeds up big updates by running it as a single rpm transaction.

@bzeller Thanks, dup finished in 22m as opposed to 30m without.
I did notice some bugs with zypper on adding it though:
1. The command `zypper --non-interactive --xmlout dist-upgrade --dry-run` starts downloading all the packages. I didn't wait to see if it would install them too!
2. Zypper's item counter had an incorrect display towards the end where it was running all the grouped together posttrans and other scripts. It showed `(2050/2050)` like everything was finished.

Seems zypper is not yet updated to fully incorporate this upstream change to libzypp.

**Comment by bzeller** [comment]
>     1. The command `zypper --non-interactive --xmlout dist-upgrade --dry-run` starts downloading all the packages. I didn't wait to see if it would install them too!
> 

I can reproduce that locally thanks, we will look into it

>     2. Zypper's item counter had an incorrect display towards the end where it was running all the grouped together posttrans and other scripts. It showed `(2050/2050)` like everything was finished.

Yes, this is a issue with how libzypp currently generates the reports... The number you see there is the number of transaction steps, previously zypper was invoking rpm per package and knew exactly how many steps there will be. By letting librpm handle the full transaction we do not know in advance how many additional steps there are, plus we have much more reports that are generated. Maybe we should simply not show any number at all for the posttrans scripts.

> Seems zypper is not yet updated to fully incorporate this upstream change to libzypp.
As said, more a issue with libzypp rather than with zypper.

**Comment by bzeller** [comment]
> > ```
> > 1. The command `zypper --non-interactive --xmlout dist-upgrade --dry-run` starts downloading all the packages. I didn't wait to see if it would install them too!
> > ```
> 
> I can reproduce that locally thanks, we will look into it

So this is actually not a bug but a requirement of librpm. In legacy mode rpm was not invoked for the dry run , but in singletrans dry run requires the packages to be locally available, because librpm needs to read the headers to calculate the dry run.

**Comment by pavinjosdev** [comment]
>> 1. The command `zypper --non-interactive --xmlout dist-upgrade --dry-run` starts downloading all the packages. I didn't wait to see if it would install them too!
> So this is actually not a bug but a requirement of librpm. In legacy mode rpm was not invoked for the dry run , but in singletrans dry run requires the packages to be locally available, because librpm needs to read the headers to calculate the dry run.

Oh, that's unexpected. Hope this behavior will be resolved if rpm single txn is made the default.
A lot of people would be surprised and complaining otherwise!
For example, zypper could say with only `--dry-run` and not --download-only that the latter is *required* to proceed. But then `--download-only` would be a *required option.* 😂

Maybe you have better ideas!

**Comment by bzeller** [comment]
To give a small update here, over the last years we did lay the groundwork to support async workflows in zypp, and have recently merged some bigger PRs that bring us closer to refreshing repositories in parallel: 

- openSUSE/libzypp#527
- openSUSE/libzypp#528

Those PRs are a complete rewrite of zypps repo refresh logic into a structure that allows us to compile it into a sync or async workflow, based on if its called from the legacy API , which we have to continue to support for now, and the new async API. The next step now is to bring up the new API which can be called from the user frontend. 

I'll keep adding updates here once we hit the next milestones.

**Comment by ell1e** [comment]
I'm not sure this was pointed out before, but on slow connections parallel downloads can cause all downloads and internet connectivity to be prone to failing entirely. Therefore, parallel downloads are only a safe default if there's some sort of detection of failing or stalling downloads that falls back to a single download at some point, or it may otherwise lead to the user being unable to download anything at all.

**Comment by Kyr4l** [comment]
> I'm not sure this was pointed out before, but on slow connections parallel downloads can cause all downloads and internet connectivity to be prone to failing entirely. Therefore, parallel downloads are only a safe default if there's some sort of detection of failing or stalling downloads that falls back to a single download at some point, or it may otherwise lead to the user being unable to download anything at all.

that's a reasonable concern, however it shouldn't be that hard to circumvent that issue :
either let users pick a max amount of parallel downloads (thru config files or with an argument to override it)
OR
make the PM check each package signature after downloading it, if a single check fails, the PM shall reduce the amount of parallel downloads automatically until the signature check succeeds.

that could be a viable solution especially for mobile devices such as laptops which often operate on many different networks with variable speeds.

**Comment by ell1e** [comment]
I think ideally, zypper would only switch on parallel downloads on a per-command basis only once the downloads surpassed some known usable minimum speed indicating a faster connection. Because if you don't manage to get the expected throughput of the mirror or even a fraction of it, I'm pretty sure parallel downloads will almost never provide any speed up over just letting tcp/ip max out a single connection as best as possible.

A config value seems insufficient: e.g. KDE Discover's update button is completely unusable to me because it uses parallel downloads with way beyond 3+ with no obvious way to disable it. Even just 3 at once often causes disaster since other programs may also be sending data. Maybe there's a way to fix KDE Discover here, but the result as a user is rather that I just don't use it.

**Comment by bzeller** [comment]
> I think ideally, zypper would only switch on parallel downloads on a per-command basis only once the downloads surpassed some known usable minimum speed indicating a faster connection. Because if you don't manage to get the expected throughput of the mirror or even a fraction of it, I'm pretty sure parallel downloads will almost never provide any speed up over just letting tcp/ip max out a single connection as best as possible.
> 
We are not going to optimize for the less common bad case in the beginning, but all downloads will be handled by a central scheduler. So in that part of the code we could try to be smart about the downloads once we have a fully working async zypper version. But a simple CLI switch could also help with that problem,  like `zypper --max-parallel-downloads=1 install world`

Also its still up for debate if zypper itself will be async at some point, or if we will have a new tool called `zypp`  to replace zypper in the long run. Too many scripts rely on the zypper behavior as it is.

**Comment by bzeller** [comment]
As promised, we have another piece of the puzzle currently pending for merge:
- openSUSE/libzypp#538

Making the RepoManager rewrite feature complete. Next steps will be the new APIs so 
that client code can trigger the async behaviors ( there is no way to do that atm aside from test cases )

**Comment by ell1e** [comment]
> We are not going to optimize for the less common bad case in the beginning

For what it's worth, depending on the machine it's not going to be less common but instead always. So the user impact of this can be pretty huge. I'm pretty sure e.g. all major digital video game distribution platforms download via a single connection by default as well, which might be related.

**Comment by smabbasht** [comment]
Can we expect `zypper` to behave like `pacman` with parallel downloads once it gets done with the update? Would it work out of the box without the need of additional CLI tools as proposed in the previous discussion?

**Comment by lietu** [comment]
I have to express my astonishment that when I asked zypper to install 500 packages it wasn't downloading them in parallel, and that when I looked into how to enable it I found this decade old feature request. This sure reinforces the appearance that openSUSE is just as dead as the availability of basic things in the package repositories make it seem.

**Comment by Kyr4l** [comment]
> I have to express my astonishment that when I asked zypper to install 500 packages it wasn't downloading them in parallel, and that when I looked into how to enable it I found this decade old feature request. This sure reinforces the appearance that openSUSE is just as dead as the availability of basic things in the package repositories make it seem.

`apt` doesn't do parallel download, does that means Debian, Ubuntu, and all their derivatives are dead ? no, it doesn't

**Comment by bzeller** [comment]
> I have to express my astonishment that when I asked zypper to install 500 packages it wasn't downloading them in parallel, and that when I looked into how to enable it I found this decade old feature request. This sure reinforces the appearance that openSUSE is just as dead as the availability of basic things in the package repositories make it seem.

Feel free to help and contribute, we are open for pull requests. Open Source lives from volunteers and different people helping out. Sadly people complaining about the current state are getting more and more and people willing to help out are getting less. We are continuously working on zypp but with limited resources and need to have backwards compatibility in mind that goes back into years old distribution releases.

**Comment by oshanz** [comment]
> > I have to express my astonishment that when I asked zypper to install 500 packages it wasn't downloading them in parallel, and that when I looked into how to enable it I found this decade old feature request. This sure reinforces the appearance that openSUSE is just as dead as the availability of basic things in the package repositories make it seem.
> 
> `apt` doesn't do parallel download, does that means Debian, Ubuntu, and all their derivatives are dead ? no, it doesn't


First, Suse is not going to die because there is no parallel download.
Yet Ubuntu and Debian are not good examples, because last time I checked, they didn't serve GB size updates often. Another rolling distribution would be a more fitting comparison.

**Comment by Firstyear** [comment]
The other thing is that throwing around vague terms like "no parallel downloads" means nothing. Network and download optimisation requires evidence and understanding of the problem space. Having sat down and spent many hours with zypper and wireshark, I can tell you now **the issue is not the lack of parallel downloads**. 

If you have a constructive report, please provide a wireshack/tcpdump output and analysis of the network traffic as well as your geographic region (US, EU, AU etc. )

**Comment by pavinjosdev** [comment]
I use Debian on servers and openSUSE Tumbleweed-Slowroll on my personal machines. APT doesn't have parallel downloads either but major version updates are quite fast and utilizes the available bandwidth to a high degree since the repo metadata and packages are all served from a CDN `deb.debian.org`. So in this light, it could be argued that zypper is not at fault here and to improve download speeds, the existing `cdn.opensuse.org` must be improved to directly handle downloads rather than acting as a redirector to a local mirror.

Frankly the alternative option of parallel downloads discussed here and as I have implemented in [zypperoni](https://github.com/pavinjosdev/zypperoni) is less user friendly because even though it helps greatly reduce latency consideration and helps in utilization of available bandwidth, the local mirror being redirected to wouldn't always be capable of such high throughput. Personally I'm using `mirrorcache-eu.opensuse.org` currently together with zypperoni as it provides me with the best possible throughput. From my earlier experiments with Arch based distros such as EndeavourOS, they had a neat utility to rank mirrors and then use that for pacman's parallel downloads. pacman in its default configuration or using the default mirrors also isn't very fast. If zypper implements parallel downloads, then a mirror ranking feature as well as max number of parallel downloads based on available bandwidth should also be added, otherwise it would have the same problems which would require users to use external tools to solve.

**Comment by Firstyear** [comment]
> the existing `cdn.opensuse.org` must be improved to directly handle downloads rather than acting as a redirector to a local mirror.

cdn.o.o doesn't act as a redirector, it's a true cdn provided by fastly. The redirectors are download.opensuse.org and the mirrorcache-<geo>.o.o and friends. 

> From my earlier experiments with Arch based distros such as EndeavourOS, they had a neat utility to rank mirrors and then use that for pacman's parallel downloads.

This is what zypper does internally, it gets a mirrorlist and then attempts to get the package from one of them, favouring closer mirrors etc. 

And this is why I keep saying "you need to understand the problem" because a lot of people here keep prescribing solutions without knowing what is actually going on. Spend a few hours with wireshark and observe the behaviour, and you see there are 3 major flaws that remain to this day. 

1) It does not properly connection pool, so when downloading from a mirror it is continually opening and closing connections which is adding latency. This can be easily seen in wireshark. 

2) zypper supports *partial repos*. This means you can't just pick one mirror and get all your packages from it. Each *package itself* requires a metadata lookup which adds latency. In otherwords, instead of a repo sending back "here are the list of mirrors you can use", zypper has to continually query "which repos have this *package from this repo*. 

3) zypper attempts to get packages in *stripes* using range requests. This means that rather than download 20MB from one mirror, it attempts to get smaller chunks of the package from 5 mirrors or more at the *same time* in parallel.  The problem is that because of TCP slow start, this means none of the connections ever build up to maximum line rate before the chunk is complete and the connection resets. This is the singular cause of issues in zypper, and I have been saying it for years now. 


Problem 3 is partially addressed by my work to increase the chunk size from I think 128kb to 4Mb. Though honestly, that was a wild guess, and it should probably be way larger. Overnight people noted that updates doubled in performance, one user reporting they went from 8 hours to 1 hour to update. 

Problem 2 is probably the most important to resolve next because almost all mirrors are whole repos, not partial, and so the need for this is reduced. If anything, if there is a need for partial mirrors, people should run https://github.com/Firstyear/opensuse-proxy-cache/ instead which can allow hot-item caching. This would remove the extra latency induced from going to a redirector for *every package* instead allowing the client to pull a mirrorlist *once* and then just route requests directly as required. 

Problem 1 is something that may or may not have been addressed, but diminishes in severity with the changes noted above. 

So again - don't prescribe solutions without understanding the problem. I've spent a lot of time to understand the problem *and implement fixes*. Parallel downloads is *not what is holding zypper back today*.

**Comment by bzeller** [comment]
>     1. It does not properly connection pool, so when downloading from a mirror it is continually opening and closing connections which is adding latency. This can be easily seen in wireshark.

Yes this is a problem, I tried to fix it but it is closely connected to how libCurl internally does connection pooling and how
the whole media backend is made in libzypp. Also related to point 2, since we download package 1 from mirror A and package 2 from mirror 2 , etc etc, at some point curl starts to cycle the connection pool... Only solution would be to do our own connection pooling of n connections  and then try to fetch from those first only making new ones when we  cannot find one in the received mirror list.


> 
>     2. zypper supports _partial repos_. This means you can't just pick one mirror and get all your packages from it. Each _package itself_ requires a metadata lookup which adds latency. In otherwords, instead of a repo sending back "here are the list of mirrors you can use", zypper has to continually query "which repos have this _package from this repo_.

Correct, this basically is the reason for not reusing connections in the way we could... as described above.

> 
>     3. zypper attempts to get packages in _stripes_ using range requests. This means that rather than download 20MB from one mirror, it attempts to get smaller chunks of the package from 5 mirrors or more at the _same time_ in parallel.  The problem is that because of TCP slow start, this means none of the connections ever build up to maximum line rate before the chunk is complete and the connection resets. This is the singular cause of issues in zypper, and I have been saying it for years now.

Yes the backend we have in development atm tries to solve all those... if it solves them right we will see as soon as I have the API at a point where I finally can implement a refresh repo tool that utilizes it.

**Comment by Firstyear** [comment]
@bzeller Thanks for that answer mate, and I hope that my comment and yours serves as an example for other commenters in this thread about evidence based approaches.

**Comment by pavinjosdev** [comment]
> cdn.o.o doesn't act as a redirector, it's a true cdn provided by fastly. The redirectors are download.opensuse.org and the mirrorcache-.o.o and friends.

Try to download an RPM from it and the file is served by a local mirror and not the CDN itself. 🥲

**Comment by lkocman** [comment]
This got again spotlight after recent https://www.reddit.com/r/openSUSE/comments/1f07gqe/tried_coming_back_to_opensuse_but_it_was_a_failure/ I made a jira to propose to enable feature for 16.

**Comment by Firstyear** [comment]
@lkocman The thing is if you actually dig into why zypper is slow, the answer is NOT lack of parallel downloads. It's "everything else" but that requires people to actually crack out wireshark or a profiler and simply look.

**Comment by mlandres** [comment]
@wcomegys but keep in mind that for zypp the content of /var/cache/zypp is expected to be safe. Packages in there are expected to have passed all configured digest and signature checks. Be sure not to put broken or insecure packages into the cache.

**Comment by wcomegys** [comment]
@mlandres I get the package list from zypper doing a zypper refresh and then a zypper list-updates, after I download each rpm package I do a rpm -checksig if it does not pass I remove it. Followed by a md5sum, sha1sum and a sha256sum. I toss out any repo strings that do not have tumbleweed or leap and the same version number for the version of Opensuse. 
I am open to any ideas to improve. This is all in my documentation.

**Comment by amirsina-mashayekh** [comment]
8 years after this feature was rqeuested and it is not implemented yet...
Today internet bandwidth is much more, and single download cannot use its full potential.
`zypper dup` on Tumbleweed downloads many small packages, usually http latency is more than download time.
I really like openSUSE but I sometimes miss `apt` which could download packages in parallel and showed an ETA.

**Comment by Firstyear** [comment]
> 8 years after this feature was rqeuested and it is not implemented yet... Today internet bandwidth is much more, and single download cannot use its full potential. `zypper dup` on Tumbleweed downloads many small packages, usually http latency is more than download time. I really like openSUSE but I sometimes miss `apt` which could download packages in parallel and showed an ETA.

https://github.com/openSUSE/zypper/issues/104#issuecomment-2311315034

Please read the thread.

**Comment by pavinjosdev** [comment]
@Firstyear Politely disagree, if this were the case then programs that do parallel downloads wouldn't have made any difference in practice but it [does](https://github.com/pavinjosdev/zypperoni?tab=readme-ov-file#performance-tests) when connected to a mirror with enough bandwidth. 🐎

@amirsina-mashayekh apt doesn't have parallel downloads but it could appear fast because:
1. Updates are few in number
2. All packages are actually served from a CDN unlike Tumbleweed where only the metadata is hosted on a CDN. 🥹

**Comment by Firstyear** [comment]
> [@Firstyear](https://github.com/Firstyear) Politely disagree, if this were the case then programs that do parallel downloads wouldn't have made any difference in practice but it [does](https://github.com/pavinjosdev/zypperoni?tab=readme-ov-file#performance-tests) when connected to a mirror with enough bandwidth. 🐎

Great, but zypperoni side steps all the *actual issues in zypper* like how it connects/disconnects rapidly, does chunked dls, and more.

So you aren't comparing the same thing. The issues with zypper today aren't the lack of parallel downloads, it's the bad connection management. 

> 
> [@amirsina-mashayekh](https://github.com/amirsina-mashayekh) apt doesn't have parallel downloads but it could appear fast because:
> 
>     1. Updates are few in number
> 
>     2. All packages are actually served from a CDN unlike Tumbleweed where only the metadata is hosted on a CDN. 🥹

This statement isn't correct. Metadata are hosted on the same mirrors as packages.

**Comment by pavinjosdev** [comment]
> Metadata are hosted on the same mirrors as packages.

I stand corrected, things have changed at cdn.o.o for the better since I last tested. 🚀

**Comment by dicer** [comment]
I can confirm that `ZYPP_PCK_PRELOAD=1 zypper dup` speeds up the download considerably! Thanks!
If there is a problem when downloading some files though, the whole process gets stopped and you have to download all files again. Would be nice to cache those downloaded files.

**Comment by bzeller** [comment]
> I can confirm that `ZYPP_PCK_PRELOAD=1 zypper dup` speeds up the download considerably! Thanks! If there is a problem when downloading some files though, the whole process gets stopped and you have to download all files again. Would be nice to cache those downloaded files.

Actually successfully downloaded files are already cached, but from the UI does not make a difference between downloaded or found in cache.

**Comment by as-muncher** [comment]
There's still a problem, though, when files don't get downloaded properly.  I do "zypper dup --download-only", and it starts to preload/download.  Then, I start qbittorrent and Session, and so eventually, zypper gets through 1.5 GB or more of downloading, just to finally say that some files weren't downloaded properly or some error like that, and then I have to run the command all over again, instead of zypper retrying the errored downloads.  To save time and to prevent that kind of nonsense, I find I have to use zypperoni.

**Comment by Mirppc** [comment]
> There's still a problem, though, when files don't get downloaded properly. I do "zypper dup --download-only", and it starts to preload/download. Then, I start qbittorrent and Session, and so eventually, zypper gets through 1.5 GB or more of downloading, just to finally say that some files weren't downloaded properly or some error like that, and then I have to run the command all over again, instead of zypper retrying the errored downloads. To save time and to prevent that kind of nonsense, I find I have to use zypperoni.

I have found a similar issue when using `env ZYPP_PCK_PRELOAD=1 zypper dup` where if a package fails to download it doesnt retry and just fails.  In further testing, or as one could refer to whacking one's head agianst the wall to see if things change, when a package fails to download initialy they will continue to fail to download in further attempts.

so for example is qt6-linguist-devel fails to download, it will continue to fail to download in all attempts using `env ZYPP_PCK_PRELOAD=1 zypper dup` .

However if one goes back to the traditional method, it will suddenly and mysteriously download with no issues at all... other than being slower.

**Comment by 13werwolf13** [comment]
> > There's still a problem, though, when files don't get downloaded properly. I do "zypper dup --download-only", and it starts to preload/download. Then, I start qbittorrent and Session, and so eventually, zypper gets through 1.5 GB or more of downloading, just to finally say that some files weren't downloaded properly or some error like that, and then I have to run the command all over again, instead of zypper retrying the errored downloads. To save time and to prevent that kind of nonsense, I find I have to use zypperoni.
> 
> I have found a similar issue when using `env ZYPP_PCK_PRELOAD=1 zypper dup` where if a package fails to download it doesnt retry and just fails. In further testing, or as one could refer to whacking one's head agianst the wall to see if things change, when a package fails to download initialy they will continue to fail to download in further attempts.
> 
> so for example is qt6-linguist-devel fails to download, it will continue to fail to download in all attempts using `env ZYPP_PCK_PRELOAD=1 zypper dup` .
> 
> However if one goes back to the traditional method, it will suddenly and mysteriously download with no issues at all... other than being slower.

I confirm, I observe exactly the same behavior. Sometimes it helps between dup attempts to make `zypper ref -fbds`, but this is more likely a coincidence than 100% working advice.

**Comment by as-muncher** [comment]
@13werwolf13 what does ref -fbds do?

**Comment by Mirppc** [comment]
> > [@13werwolf13](https://github.com/13werwolf13) what does ref -fbds do?
> 
> ```
> [werwolf@home] ~/
> ❯ LANG=c zypper ref -h
> refresh (ref) [ALIAS|#|URI] ...
> 
> Refresh repositories specified by their alias, number or URI. If none are specified, all enabled
> repositories will be refreshed.
> 
>   Command options:
> 
> -f, --force                 Force a complete refresh. Default: false
> -b, --force-build           Force rebuild of the database. Default: false
> -d, --force-download        Force download of raw metadata. Default: false
> -B, --build-only            Only build the database, don't download metadata. Default: false
> -D, --download-only         Only download raw metadata, don't build the database. Default: false
>     --include-all-archs     Multi-Arch repos: Download raw metadata for all offered architectures
>                             even if the repo supports filtering. Default: false
> -r, --repo <ALIAS|#|URI>    Refresh only specified repositories.
> -s, --services              Refresh also services before refreshing repos. Default: false
> ```
So far in my testing... this didnt solve the issue.

**Comment by bzeller** [comment]
> There's still a problem, though, when files don't get downloaded properly. I do "zypper dup --download-only", and it starts to preload/download. Then, I start qbittorrent and Session, and so eventually, zypper gets through 1.5 GB or more of downloading, just to finally say that some files weren't downloaded properly or some error like that, and then I have to run the command all over again, instead of zypper retrying the errored downloads. To save time and to prevent that kind of nonsense, I find I have to use zypperoni.

This is exactly the reason why its still a tech preview and not a released feature. Depending on how you did set up your repo the preloader does not have a mirrors ... ( it only has when using metalink in the .repo file ). We are polishing the feature right now and will release it with one of the upcoming updates. Once it lands it will be automatically enabled for everyone.

**Comment by Firstyear** [comment]
@bzeller Who do we need to communicate with about this? This sync issue should be solvable and should really not be an issue today in 2025 because we should be able to do this atomically. Every other distro manages except us, so we really need to raise this as a large issue.

**Comment by bcomegys** [comment]
Hi, I agree this should be able to support multiple mirrors and make sure a downloaded RPM file is valid.
I wrote zkk opensuse which uses the return from zypper to get the needed updates. zkk downloads in parallel, I think up to 42 (settable) at one time using aria2c. It runs a rpm checksig to make sure the file is good. zkk creates another repo directory in which each repo can (you have to add the URLs, if it fails it will go to the next repo) have multiple repos zkk tracks the download speed and will pick the fastest repos. zkk has many settings, excellent logging and good documentation. It is written in REXX so it is a little slower but provides many good ideas.

**Comment by as-muncher** [comment]
@bcomegys How is your program better than zypperoni?

**Comment by Mirppc** [comment]
> [@bcomegys](https://github.com/bcomegys) How is your program better than zypperoni?

you have to install zypperoni.  Zypper is the stock default package management tool.

**Comment by bzeller** [comment]
> [@bzeller](https://github.com/bzeller) Who do we need to communicate with about this? This sync issue should be solvable and should really not be an issue today in 2025 because we should be able to do this atomically. Every other distro manages except us, so we really need to raise this as a large issue.

There is no  need to talk with anyone, as I said we are polishing the feature and it will use the mirrors that are specified in the repo file ( via a mirror list in baseUrl, a metalink file or a mirrorlist file ) and to support existing setups the code will auto detect mirrors for opensuse servers.

**Comment by Firstyear** [comment]
No, there is a problem @bzeller - the data that is synced to mirrors is out of order, and is removing files when they are still needed, or syncs them out of order. We see it really commonly here where refreshes or updates don't work because mirrors aren't done syncing (they shipped metadata before the packages were rsynced), and sometimes they also remove packages prematurely when people are still using them. 

Currently the iissue is from what I can see externally is that the mirror sync assumes the mirror is just an rsync clone, when it needs to actually be A/B snapshots of the current data and data minus 1. 

The process should be:

* Sync new packages (version X)
* Sync new metadata (version X)
* Delete old metadata (vervsion X minus 2)
* Delete old packages (version X minus 2)

That way the mirror has both X and X minus 1 so clients that refreshed earlier still get the older content, and metadata isn't updated until all the packages are available and stable.

But again - today this isn't happening and it's why we see issues like the above because currently the mirror sync is pushing out metadata to soon, or it's deleting packages that are still needed and the new packages are pushed, but the new metadata isn't ready. It's a mess.

So it is an issue, and it needs to be brought up. Who do we bring it up to?

**Comment by as-muncher** [comment]
> > [@bcomegys](https://github.com/bcomegys) How is your program better than zypperoni?
> 
> you have to install zypperoni. Zypper is the stock default package management tool.

No.  zkk opensuse is what I meant

**Comment by bzeller** [comment]
> No, there is a problem [@bzeller](https://github.com/bzeller) - the data that is synced to mirrors is out of order, and is removing files when they are still needed, or syncs them out of order. We see it really commonly here where refreshes or updates don't work because mirrors aren't done syncing (they shipped metadata before the packages were rsynced), and sometimes they also remove packages prematurely when people are still using them.

Yes I know, we brought this up many times, but the answer is always: We can not control  how the mirrors are synced, 
so we cannot change it.

> 
> Currently the iissue is from what I can see externally is that the mirror sync assumes the mirror is just an rsync clone, when it needs to actually be A/B snapshots of the current data and data minus 1.
> 
> The process should be:
> 
>     * Sync new packages (version X)
> 
>     * Sync new metadata (version X)
> 
>     * Delete old metadata (vervsion X minus 2)
> 
>     * Delete old packages (version X minus 2)
>    
Yes Michael suggested  that many times already. But how do you enforce this? 

> So it is an issue, and it needs to be brought up. Who do we bring it up to?

I usually talk to @andrii-suse but he also does not have control over the mirrors.  

Not sure what would be the appropriate place to discuss this topic. Here none of the people who actually can influence it are noticed.

**Comment by andrii-suse** [comment]
I agree that zypper is not right project for discussing how packages are delivered in openSUSE infrastructure. I believe there are few places that mention this. I was trying to work on it in https://progress.opensuse.org/issues/134702 , but it really needs some dedication (and I have better considerations since when I tried to document the ideas in that ticket).

**Comment by Firstyear** [comment]
But also *there is* something that can be done - OBS when it builds packages and submits them to the master mirror, can do the A/B stage and ensure that two versions exist at a time, and it can also hold back the metadata until everything else is done.

I'm also pretty sure every other distro manages this so, so we have to find a way to tell mirror owners to update their sync tool. 

So is this something we bring up on progress.o.o? Or OBS?

**Comment by wcomegys** [comment]
Hi. I think I tried zypperoni once, a long time ago. I think zypperoni is
much faster, as a person who envisions a product we decided to emphasize
certain ideas and goals, I do not know all the ideas the zypperoni author
decided to emphasize, I think one was speed and parallel downloads, which
he accomplished very well.
I emphasized logging, I always develop my logging section first. I know my
program would be much faster if I was to toss out the logging and stats
collection. zkk pulls this information from the aria2 download stats and
periodically updates the zkk repo file URL stats.

zkk does check and add to the log file each rpm file it downloads it first
does a rpm -checksig and a sha-256, sha-1 and md5sum hash so you know the
file you downloaded including URL and you can track it back if you need
to, here is an example:
20241115 22:08:36.842389 15505        ZK_downLoader
         download.opensuse.org-oss_1 ZKK_Repo_Stem.1 is 1 0000000001997291
0000333 0000000 http://download.opensuse.org/tumbleweed/repo/oss
20241115 22:08:37.438583 15505        ZK_downLoader
         download.opensuse.org-oss_1 Aria2 Download Results:
20241115 22:08:37.484735 15505        ZK_downLoader
         download.opensuse.org-oss_1 Aria2 301c87|OK  |
 n/a|/var/cache/zypp/packages/download.opensuse.org-oss_1/x86_64/plasma6-workspace-libs-6.2.3-3.1.x86_64.rpm
20241115 22:08:37.528061 15505        ZK_downLoader
         download.opensuse.org-oss_1 Aria2 8c6e93|OK  |
2.4MiB/s|/var/cache/zypp/packages/download.opensuse.org-oss_1/x86_64/plasma6-workspace-libs-6.2.3-3.1.x86_64.rpm
20241115 22:08:37.575930 15505        ZK_downLoader
         download.opensuse.org-oss_1 Aria2 (OK):download completed.
20241115 22:08:37.634075 15505        ZK_downLoader
         download.opensuse.org-oss_1 * Information Exiting * File:
/var/cache/zypp/packages/download.opensuse.org-oss_1/x86_64/plasma6-workspace-libs-6.2.3-3.1.x86_64.rpm,
digests signatures OK, SHA-256 Hash:
d8ffc3f3d3fcdcbb50d44b29f5c91baebc57a2e5b001f8672f8384f1dec15bc7, SHA-1
Hash: 8c6778a252ee3fc286971ca57713c77351fdf88d, MD5 Hash:
efe109692521fa2b153b818e658f29cf

Above the number 15505 is the PID, Here the logging thread will take the
information 2.4MiB/s and modify the average download speed
0000000001997291  and add another successful download to 0000333 = 334 and
on exit or before update the zkk repo file.
If the rpm checksig did not pass zkk would delete the file, download it or
re-download it. EVEN IF IT DID NOT DOWNLOAD THE FILE

I wanted to experiment with repo files which support multiple download
repos with speed collection and sorting so the fastest ones are first and
you can choose to use the top N, with automatic failover, one can be down
and if you have multiple repos zkk will just failover to the next one.
1 0000000001998707 0002011 0000000
http://download.opensuse.org/tumbleweed/repo/oss
Enabled   Ave Speed   Good Down  Fail    URL
   You can have as many repos locations as you want, if you do not like a
repo location you can go in and disable it by changing the first 1 to a 0
(zero), provided you have another location for that repo, the program also
tracks failures and adds date and timestamps for those, all described in
the documentation. The fastest enabled ones will always be at the     top
of the choice.
I think zkk has good configuration through its config file, I do not modify
any opensuse files, I build parallel ones.
I think zkk has good documentation, you can read it here:
https://github.com/wcomegys/zkk.opensuse/blob/main/zkk.pdf

So people are experimenting with different ideas, the better ones will get
incorporated like parallel downloading, I would hope checking the rpm file
is valid :) (I read a comment to this up above), and a checksum and adding
the information into a log, it should be running on a thread it's not like
it is going to bog down the CPU. Maybe keeping track of performance data
and being able to automatically use the top N fastest repos in your area.
I hope this answers your question, and again here is the link to the
documentation:
https://github.com/wcomegys/zkk.opensuse/blob/main/zkk.pdf

**Comment by Mirppc** [comment]
> Hi. I think I tried zypperoni once, a long time ago. I think zypperoni is
> much faster, as a person who envisions a product we decided to emphasize
> certain ideas and goals, I do not know all the ideas the zypperoni author
> decided to emphasize, I think one was speed and parallel downloads, which
> he accomplished very well.
> I emphasized logging, I always develop my logging section first. I know my
> program would be much faster if I was to toss out the logging and stats
> collection. zkk pulls this information from the aria2 download stats and
> periodically updates the zkk repo file URL stats.
> 
> zkk does check and add to the log file each rpm file it downloads it first
> does a rpm -checksig and a sha-256, sha-1 and md5sum hash so you know the
> file you downloaded including URL and you can track it back if you need
> to, here is an example:
> 20241115 22:08:36.842389 15505        ZK_downLoader
>          download.opensuse.org-oss_1 ZKK_Repo_Stem.1 is 1 0000000001997291
> 0000333 0000000 http://download.opensuse.org/tumbleweed/repo/oss
> 20241115 22:08:37.438583 15505        ZK_downLoader
>          download.opensuse.org-oss_1 Aria2 Download Results:
> 20241115 22:08:37.484735 15505        ZK_downLoader
>          download.opensuse.org-oss_1 Aria2 301c87|OK  |
>  n/a|/var/cache/zypp/packages/download.opensuse.org-oss_1/x86_64/plasma6-workspace-libs-6.2.3-3.1.x86_64.rpm
> 20241115 22:08:37.528061 15505        ZK_downLoader
>          download.opensuse.org-oss_1 Aria2 8c6e93|OK  |
> 2.4MiB/s|/var/cache/zypp/packages/download.opensuse.org-oss_1/x86_64/plasma6-workspace-libs-6.2.3-3.1.x86_64.rpm
> 20241115 22:08:37.575930 15505        ZK_downLoader
>          download.opensuse.org-oss_1 Aria2 (OK):download completed.
> 20241115 22:08:37.634075 15505        ZK_downLoader
>          download.opensuse.org-oss_1 * Information Exiting * File:
> /var/cache/zypp/packages/download.opensuse.org-oss_1/x86_64/plasma6-workspace-libs-6.2.3-3.1.x86_64.rpm,
> digests signatures OK, SHA-256 Hash:
> d8ffc3f3d3fcdcbb50d44b29f5c91baebc57a2e5b001f8672f8384f1dec15bc7, SHA-1
> Hash: 8c6778a252ee3fc286971ca57713c77351fdf88d, MD5 Hash:
> efe109692521fa2b153b818e658f29cf
> 
> Above the number 15505 is the PID, Here the logging thread will take the
> information 2.4MiB/s and modify the average download speed
> 0000000001997291  and add another successful download to 0000333 = 334 and
> on exit or before update the zkk repo file.
> If the rpm checksig did not pass zkk would delete the file, download it or
> re-download it. EVEN IF IT DID NOT DOWNLOAD THE FILE
> 
> I wanted to experiment with repo files which support multiple download
> repos with speed collection and sorting so the fastest ones are first and
> you can choose to use the top N, with automatic failover, one can be down
> and if you have multiple repos zkk will just failover to the next one.
> 1 0000000001998707 0002011 0000000
> http://download.opensuse.org/tumbleweed/repo/oss
> Enabled   Ave Speed   Good Down  Fail    URL
>    You can have as many repos locations as you want, if you do not like a
> repo location you can go in and disable it by changing the first 1 to a 0
> (zero), provided you have another location for that repo, the program also
> tracks failures and adds date and timestamps for those, all described in
> the documentation. The fastest enabled ones will always be at the     top
> of the choice.
> I think zkk has good configuration through its config file, I do not modify
> any opensuse files, I build parallel ones.
> I think zkk has good documentation, you can read it here:
> https://github.com/wcomegys/zkk.opensuse/blob/main/zkk.pdf
> 
> So people are experimenting with different ideas, the better ones will get
> incorporated like parallel downloading, I would hope checking the rpm file
> is valid :) (I read a comment to this up above), and a checksum and adding
> the information into a log, it should be running on a thread it's not like
> it is going to bog down the CPU. Maybe keeping track of performance data
> and being able to automatically use the top N fastest repos in your area.
> I hope this answers your question, and again here is the link to the
> documentation:
> https://github.com/wcomegys/zkk.opensuse/blob/main/zkk.pdf
> […](#)

This is just muddying this support.  This is about integrated zypper parellization, which both ZKK and Zypperoni still call upon.  The problem  talked about is repo mirror level and will effect both.

@Firstyea @bzeller would it be better if a new issue where opened regarding the toping of zypper's repo issue?  Since it seems that is a separate issue.   I understand that discussion about it will be over on the Open Build Service Mailing List and possibly project.o.o.  However having something here on git that can also be followed might be useful, whether it is an issue or a discussion.

**Comment by bzeller** [comment]
> This is just muddying this support. This is about integrated zypper parellization, which both ZKK and Zypperoni still call upon. The problem talked about is repo mirror level and will effect both.
> 
> @Firstyea [@bzeller](https://github.com/bzeller) would it be better if a new issue where opened regarding the toping of zypper's repo issue? Since it seems that is a separate issue. I understand that discussion about it will be over on the Open Build Service Mailing List and possibly project.o.o. However having something here on git that can also be followed might be useful, whether it is an issue or a discussion.

Not as a zypper bug, as I said before this is completely out of the control of the zypper team and has nothing to do with zypper in the first place. 
 
@wcomegys Please stop using this bug to discuss about solutions like zkk or zypperoni this is the zypper bug tracker.

**Comment by Patrick-Poitras** [comment]
Back to OpenSUSE after an absence of a few months. The new speed is amazing. Thanks all.
